### ISSUE TITLE
Upcoming `lu` issue on nightly

### ISSUE THREAD

**Issue by dkarrasch** [JuliaDiff/TaylorSeries.jl#281]
On current master, there is a finer distinction in `copy_oftype`-like constructors, which will likely make this package fail on nightly, see

https://github.com/JuliaLang/julia/pull/40831#issuecomment-864385804

This affects your own `lu` implementation, which uses `copy_oftype(A, S)`, which used to call `convert(AbstractArray{S}(A))`, but now relies on 2-arg `similar`. So there are two options to fix this:

1. Write a 2-arg `similar` function for matrices with eltype `Taylor1`, then `copy_oftype` will start using this starting from Julia v1.8 "silently".
2. Replace `copy_oftype(A, S)` in `lu` by a direct call to `convert(AbstractArray{S}, A)`, which is handled internally in your package.

If needed, I'd be willing to help, probably just like @daanhb.

**Comment by lbenet** [comment]
Hi @dkarrasch ! Thanks for reporting the future issue. If you can, we would be grateful if you can help us with this.

**Comment by daanhb** [comment]
Hi, I'm happy to chime in. From the surrounding comments in the code, it seems the goal is to avoid pivoting, which LinearAlgebra would otherwise do by default but it may fail in this case. If that is so, might the following be sufficient?
```julia
lu(A::AbstractMatrix{Taylor1{T}}; check::Bool = true) where {T<:Number} = lu(A, Val(false); bool=bool)
```

This is not quite the same as the current implementation though, because the code tries again with pivoting if the first attempt fails. So perhaps
```julia
function lu(A::AbstractMatrix{Taylor1{T}}; check::Bool = true) where {T<:Number}
    F = lu(A, Val(false); bool=bool)
    if issuccess(F)
        return F
    else
        lu(A, Val(true); bool=bool)
    end
end
````

The advantage of the above is that you avoid using LinearAlgebra internals, but perhaps I'm missing something. If it turns out that LinearAlgebra doesn't properly copy the array, say, then we'll have to look into that.

**Comment by dkarrasch** [comment]
I tested a little bit and found that the issue is the new `lu(::StridedMatrix)` method, which takes precedence over the `lu(::AbstractMatrix{Taylor1})` method defined here. In #41288, I have removed that method assuming (https://github.com/JuliaLang/julia/pull/40831#issuecomment-863261577) that `copy_toarray` and `copy_oftype` is no different on `StridedMatrix`es.

**Comment by lbenet** [comment]
Thanks @daanhb for chiming in! I guess by now you have already found that the changes implemented in #223 were motivated by a somewhat similar report #221. I'm not an expert in Linear Algebra, so any help from you is welcome.

As a side question, do you know if we can test now 1.7 and 1.8 using github actions?

**Comment by dkarrasch** [comment]
For 1.7, you would need to "hardcode" it as "1.7". Version 1.8 is available as "nightly", just "1" corresponds to v1.6.1 currently.

**Comment by lbenet** [comment]
I just tested (locally) Julia v1.7.0-beta2 and got the following warning:
```julia
Tests for Taylor1 expansions |  375    375
┌ Warning: `lu!(A::Union{StridedMatrix, HermOrSym, Tridiagonal}, ::Val{false}; check::Bool = true)` is deprecated, use `lu!(A, NoPivot(); check = check)` instead.
│   caller = ip:0x0
└ @ Core :-1
```
Aside from this warning, tests pass fine.

**Comment by dkarrasch** [comment]
Yes, the "problematic" PR is not included in v1.7, it's going to be in v1.8, which is why tests will fail now on nightly. But https://github.com/JuliaLang/julia/pull/41288/ is going to fix that issue. There is WIP for the deprecation warning, PR is coming soon.
